#354: Adapter notes must be strings (#385)
exasol/virtual-schemas: doc/development/api/virtual_schema_api.md

### Schema Metadata

> This document is usually embedded in responses from the Adapter and informs the database about all metadata of the Virtual Schema, especially the contained Virtual Tables and it's columns.

Author: ```suggestion
This document is usually embedded in responses from the Adapter and informs the database about all metadata of the Virtual Schema, especially the contained Virtual Tables and its columns.
```
Fixed.
Can I also introduce you to [GitHub's suggestion function](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/commenting-on-a-pull-request#adding-line-comments-to-a-pull-request)? Makes reviews a lot more convenient for both author and reviewers.

> The Adapter can store so called `adapterNotes` on each level (schema, table, column), to remember information which might be relevant for the Adapter in future. Adapter note are simple strings. You can serialize objects into those strings of course, but keep in mind that the strings are embedded inside the Virtual Schemas JSON protocol, which makes quoting of conflicting characters necessary.

Author: ```suggestion
The Adapter can optionally store so called `adapterNotes` on each level (schema, table, column), to remember information which might be relevant for the Adapter in future. Adapter notes are simple strings. You can serialize objects into those strings of course, but keep in mind that the strings are embedded inside the Virtual Schemas JSON protocol, which makes quoting of conflicting characters necessary.
```
Fixed.

Some options to deal with the embedding issue:

1. After serialization use Base64 encoding or
1. Use a serialization that does not have conflicting characters like a simple CSV format or
1. Quote conflicting characters

Which variant you should choose depends on considerations like ammount of data to be transmitted, original data format and encoding overhead.

> In the example below, the Adapter remembers the table partitioning and the data type of a column which is not directly supported in EXASOL. The Adapter has these information during pushdown and can consider the table partitioning during pushdown or can add an appropriate cast for the column.

Reviewer: Pre-existing typo: Replace "these information" with "this information".

Please change the example values to strings (1 example above, 3 below). I would just use the quoting approach here.

Author: You are right. I just learned [something new about the English language](https://jakubmarian.com/information-is-vs-informations-are-is-information-plural-or-singular-in-english/). I thought this is a case like "the police" always being plural.

> In the example below, the Adapter remembers the table partitioning and the data type of a column which is not directly supported in EXASOL. The Adapter has these information during pushdown and can consider the table partitioning during pushdown or can add an appropriate cast for the column.

Author: ```suggestion
In the example below, the Adapter remembers the table partitioning and the data type of a column which is not directly supported in Exasol. The Adapter has this information during push-down and can consider the table partitioning during push-down or can add an appropriate cast for the column.
```

Some options to deal with the embedding issue:

1. After serialization use Base64 encoding or
1. Use a serialization that does not have conflicting characters like a simple CSV or key-value format or
1. Quote conflicting characters

Which variant you should choose depends on considerations like amount of data to be transmitted, original data format and encoding overhead.

In the example below, the Adapter remembers the table partitioning and the data type of a column which is not directly supported in Exasol. The Adapter has this information during push-down and can consider the table partitioning during push-down or can add an appropriate cast for the column.

> This example also demonstrates serialization in adapter notes via key-value encoding. As mentioned above more sophisticated serializations are possible as long as you make sure adapter notes are a valid string inside the JSON format by encoding or quoting.

Reviewer: Not sure about this one:
```suggestion
This example also demonstrates serialization in adapter notes via key-value encoding. As mentioned above more sophisticated serializations are possible as long as you make sure adapter notes are a valid string in the JSON format by encoding or quoting.
```

```json
{"schemaMetadata":{

>     "adapterNotes": "lastRefreshed=2015-03-01 12:10:01;anotherKey=More custom schema state here"
>     },

Reviewer: ```suggestion
    "adapterNotes": "lastRefreshed=2015-03-01 12:10:01;anotherKey=More custom schema state here",
```

    "tables": [
    {
        "type": "table",
        "name": "EXASOL_CUSTOMERS",

>         "adapterNotes": "hivePartitionColumns=CREATED,COUNTRY_ISO"
>         },

Reviewer: ```suggestion
        "adapterNotes": "hivePartitionColumns=CREATED,COUNTRY_ISO",
```

            "default": "foo",
            "isNullable": false,
            "comment": "The official name of the company",

>             "adapterNotes": "hiveDataType=List<String>"
>             }

Reviewer: ```suggestion
            "adapterNotes": "hiveDataType=List<String>"
```

}}
```

Notes

> * `adapterNotes` is an optional field which can be attached to the schema, a table or a column. It can be an arbitrarily nested Json document.

Reviewer: I would keep the word "optional" somewhere.

Author: Added above.
